BUILD.md

```diff
@@ -0,0 +1,472 @@
+to the Conduit controller pod:
+
+```bash
+POD_NAME=$(kubectl --namespace=conduit get po --selector=app=controller -o jsonpath='{.items[*].metadata.name}')
+kubectl -n conduit port-forward $POD_NAME 8085:8085
+```
+
+Then connect the local web process to the forwarded port:
+
+```bash
+bin/go-run web --api-addr=localhost:8085
+```
+
+### Webpack dev server
+
+To develop with a webpack dev server, start the server in a separate window:
+
+```bash
+cd web/app
+yarn webpack-dev-server
+```
+
+And then set the `--webpack-dev-server` flag when running the web server:
+
+```bash
+bin/go-run web --webpack-dev-server=http://localhost:8080
+```
+
+To add a JS dependency:
+
+```bash
+cd web/app
+yarn add [dep]
+```
+
+## Proxy
+
+To build and run the Rust proxy:
+
+```bash
+cargo build -p conduit-proxy
+CONDUIT_PROXY_LOG=trace \
+  CONDUIT_PROXY_PUBLIC_LISTENER=tcp://0:5432 \
+  CONDUIT_PROXY_PRIVATE_FORWARD=tcp://127.0.0.1:1234 \
+  CONDUIT_PROXY_CONTROL_URL=tcp://127.0.0.1:8086 \
+  target/debug/conduit-proxy
+```
+
+To connect to a live `proxy-api` at `localhost:8086`:
+
+```bash
+bin/go-run controller/cmd/proxy-api
+```
+
+# Testing
+
+## Go
+
+```bash
+dep ensure && dep prune
+go test ./...
+go vet ./...
+```
+
+## JavaScript
+
+```bash
+cd web/app
+yarn && yarn webpack
+yarn karma start --single-run
+```
+
+## Rust
+
+```bash
+cargo check
+cargo test
+```
+
+# Updating protobuf dependencies
+
+If you make Protobuf changes, run:
+
+```bash
+bin/protoc-go.sh
+```
+
+# Updating Docker dependencies
+
+The Rust proxy and Go Docker images rely on base dependency images with
+hard-coded SHA's:
+
+`gcr.io/runconduit/go-deps` depends on
+- [`Gopkg.lock`](Gopkg.lock)
+- [`Dockerfile-go-deps`](Dockerfile-go-deps)
+
+`gcr.io/runconduit/proxy-deps` depends on
+- [`Cargo.lock`](Cargo.lock)
+- [`proxy/Dockerfile-deps`](proxy/Dockerfile-deps)
+
+If any of these files change, update the Dockerfile SHA's with:
+
+```bash
+GO_DEPS_SHA=$(sh -c ". bin/_tag.sh && go_deps_sha")
+PROXY_DEPS_SHA=$(sh -c ". bin/_tag.sh && proxy_deps_sha")
+
+find . -type f -name 'Dockerfile*' -exec sed -i '' -e 's/gcr\.io\/runconduit\/go-deps:[^ ]*/gcr\.io\/runconduit\/go-deps:'$GO_DEPS_SHA'/g' {} \;
+find . -type f -name 'Dockerfile*' -exec sed -i '' -e 's/gcr\.io\/runconduit\/proxy-deps:[^ ]*/gcr\.io\/runconduit\/proxy-deps:'$PROXY_DEPS_SHA'/g' {} \;
+```
+
+# Go run
+
+Our Go build instructions use a [`bin/go-run`](bin/go-run) script in lieu
+`go run`. This is a convenience script that leverages caching via `go build` to
+make your build/run/debug loop faster.
+
+In general, replace commands like this:
+
+```bash
+go run web/main.go
+```
+
+with this:
+
+```bash
+bin/go-run web
+```
+
+# Build Architecture
+
+![Build Architecture](https://g.gravizo.com/source/svg/build_architecture?https%3A%2F%2Fraw.githubusercontent.com%2Frunconduit%2Fconduit%2Fsiggy%2Fbuild%2FBUILD.md)
+
+<details>
+<summary></summary>
+build_architecture
```

Comment:
Do others think that this adds clarity? To me this just says "it's complex."  I find the scripts themselves to be a bit easier approach than a big dependency graph...  If they're not approachable, perhaps they could use some better documentation internally?

Reply:
Yeah, I found the labels on here to be too small to read without straining my eyes. I think some form of graph could be useful, but all I really got out of this was "we have too many different shell scripts"...

Reply:
While not specifically necessary to build conduit, I found this graph invaluable when debugging build issues or making changes in #117 and #123. Without this graph in documentation, I'd have to reconstitute it in my head each time I context switch into working on the build.

@hawkw Agree the labels are a bit small. The image is clickable to expand. ;)